### PR TITLE
P1862R1 Ranges adaptors for non-copyable iterators

### LINKED ISSUE #3417: [2019-11 LWG Motion 15] P1862R1 Ranges adaptors for non-copyable iterators
http://wiki.edg.com/pub/Wg21belfast/StrawPolls/p1862r1.pdf

These five papers (together) resolve NB comments US272 and DE288

### PR DESCRIPTION
**Pull request by jensmaurer**
Partially addresses NB US 272 and DE 288 (C++20 CD)

Fixes #3417
Fixes cplusplus/nbballot#266
Partially addresses cplusplus/nbballot#268
Partially addresses cplusplus/nbballot#284
Fixes cplusplus/papers#616

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -1276,14 +1276,17 @@
...
+    [[nodiscard]] constexpr I begin() requires (!copyable<I>);
     constexpr S end() const;
 
     constexpr bool empty() const;
     constexpr @\placeholdernc{make-unsigned-like-t}@(iter_difference_t<I>) size() const
       requires (K == subrange_kind::sized);
 
-    [[nodiscard]] constexpr subrange next(iter_difference_t<I> n = 1) const;
+    [[nodiscard]] constexpr subrange next(iter_difference_t<I> n = 1) const &
+      requires forward_iterator<I>
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
      requires forward_iterator<I>;
```

File: source/ranges.tex [hunk 2]
```diff
@@ -1316,6 +1319,10 @@
   template<size_t N, class I, class S, subrange_kind K>
     requires (N < 2)
   constexpr auto get(const subrange<I, S, K>& r);
+
+  template<size_t N, class I, class S, subrange_kind K>
+    requires (N < 2)
+    constexpr auto get(subrange<I, S, K>&& r);
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  constexpr auto get(subrange<I, S, K>&& r);
```

File: source/ranges.tex [hunk 3]
```diff
@@ -1413,6 +1420,17 @@
 Equivalent to: \tcode{return begin_;}
 \end{itemdescr}
 
+\indexlibrarymember{begin}{subrange}%
+\begin{itemdecl}
+[[nodiscard]] constexpr I begin() const requires (!copyable<I>);
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
[[nodiscard]] constexpr I begin() requires (!copyable<I>);
```

File: source/ranges.tex [hunk 4]
```diff
@@ -3438,6 +3486,17 @@
 Equivalent to: \tcode{return current_;}
 \end{itemdescr}
 
+\indexlibrarymember{base}{transform_view::iterator}%
+\begin{itemdecl}
+constexpr iterator_t<Base> base() &&
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
constexpr iterator_t<Base> base() &&;
```

File: source/ranges.tex [hunk 5]
```diff
@@ -4625,7 +4686,10 @@
   \placeholder{OUTERC} and \placeholder{INNERC} each model
   \tcode{derived_from<for\-ward_iterator_tag>}, \tcode{iterator_category}
   denotes \tcode{forward_iterator_tag}.
-\item Otherwise, \tcode{iterator_category} denotes \tcode{input_iterator_tag}.
+\item Otherwise, if \placeholder{OUTERC} and \placeholder{INNERC} each model
+  \tcode{\libconcept{derived_from}<input_iterator_tag>},
+\tcode{iterator_category} denotes \tcode{input_iterator_tag}.
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  \tcode{iterator_category} denotes \tcode{input_iterator_tag}.
```

File: source/ranges.tex [hunk 6]
```diff
@@ -5235,6 +5300,16 @@
 Equivalent to: \tcode{return inner_iterator<Const>\{i_\};}
 \end{itemdescr}
 
+\indexlibrarymember{begin}{split_view::outer_iterator::value_type}%
+\begin{itemdecl}
+constexpr inner_iterator<Const> begin() const requires (!copyable<outer_iterator>);
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
constexpr inner_iterator<Const> begin() requires (!copyable<outer_iterator>);
```

File: source/ranges.tex [hunk 7]
```diff
@@ -5358,17 +5436,23 @@
...
 \begin{codeblock}
-auto cur = x.i_.@\placeholder{current}@;
-auto end = ranges::end(x.i_.parent_->base_);
-if (cur == end) return true;
 auto [pcur, pend] = subrange{x.i_.parent_->pattern_};
-if (pcur == pend) return x.incremented_;
-do {
+auto end = ranges::end(x.i_.parent_->base_);
+if constexpr (@\exposconcept{tiny-range}@<Pattern>) {
+  const auto & cur = x.i_.current ;
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  const auto & cur = x.i_.@\placeholder{current}@;
```

File: source/ranges.tex [hunk 8]
```diff
@@ -5358,17 +5436,23 @@
...
+  if (cur == end) return true;
+  if (pcur == pend) return x.incremented_;
+  return *cur == *pcur;
+} else {
+  auto cur = x.i_.current ;
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  auto cur = x.i_.@\placeholder{current}@;
```

File: source/ranges.tex [hunk 9]
```diff
@@ -5358,17 +5436,23 @@
...
+  if (cur == end) return true;
+  if (pcur == pend) return x.incremented_;
+  do {
   if (*cur != *pcur) return false;
   if (++pcur == pend) return true;
 } while (++cur != end);
 return false;
-\end{codeblock}
+}\end{codeblock}
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
}
\end{codeblock}
```

File: source/ranges.tex [hunk 10]
```diff
@@ -5358,17 +5436,23 @@
...
+  if (pcur == pend) return x.incremented_;
+  return *cur == *pcur;
+} else {
+  auto cur = x.i_.current ;
+  if (cur == end) return true;
+  if (pcur == pend) return x.incremented_;
+  do {
   if (*cur != *pcur) return false;
   if (++pcur == pend) return true;
 } while (++cur != end);
```
Comment (johelegp) [inline source/ranges.tex]: The closing brace of the do while needs to be align with the `do`, and its statements indented two more spaces.

File: source/ranges.tex [hunk 11]
```diff
@@ -6171,12 +6257,13 @@
...
 \effects
-Initializes \tcode{current_} with \tcode{i.current_}.
+Initializes \tcode{current_} with \tcode{std::move(i.current_)}.
 \end{itemdescr}
 
 \indexlibrarymember{base}{elements_view::iterator}%
 \begin{itemdecl}
-constexpr iterator_t<base_t> base() const;
+constexpr iterator_t<base_t> base() const &
+  requires copyable<iterator_t<V>>;
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
  requires copyable<iterator_t<base_t>>;
```
Like in the description.

File: source/ranges.tex [hunk 12]
```diff
@@ -6185,6 +6272,17 @@
 Equivalent to: \tcode{return current_;}
 \end{itemdescr}
 
+\indexlibrarymember{base}{elements_view::iterator}%
+\begin{itemdecl}
+constexpr iterator_t<base_t> base() const &&;
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
constexpr iterator_t<base_t> base() &&;
```

File: source/ranges.tex (hunk 2 above)
Comment (zygoloid) [reply source/ranges.tex]: The indentation here is correct; the previous `get` is wrong.

**Pull request merged**